packages/grid/x-data-grid/src/components/containers/GridRootStyles.ts

```diff
@@ -37,174 +45,236 @@ export const GridRootStyles = styled('div', {
+  const borderColor = getBorderColor(t);
+  const radius = t.shape.borderRadius;
+
+  const containerBackground = t.vars
+    ? t.vars.palette.background.default
+    : t.palette.background.default;
+
+  // eslint-disable-next-line no-nested-ternary
+  const pinnedBackground = t.vars
+    ? t.vars.palette.background.default // We cannot lighten/darken the background color if it's a CSS variable.
+    : t.palette.mode === 'dark'
+    ? lighten(t.palette.background.default, 0.1)
+    : darken(t.palette.background.default, 0.05);
+
+  const overlayBackground = t.vars
+    ? `rgba(${t.vars.palette.background.defaultChannel} / ${t.vars.palette.action.disabledOpacity})`
+    : alpha(t.palette.background.default, t.palette.action.disabledOpacity);
+
+  const hoverOpacity = (t.vars || t).palette.action.hoverOpacity;
+  const hoverBackground = (t.vars || t).palette.action.hover;
+
+  const selectedOpacity = (t.vars || t).palette.action.selectedOpacity;
+  const selectedBackground = t.vars
+    ? `rgba(${t.vars.palette.primary.mainChannel} / ${selectedOpacity})`
+    : alpha(t.palette.primary.main, selectedOpacity);
+
+  const selectedHoverBackground = t.vars
+    ? `rgba(${t.vars.palette.primary.mainChannel} / calc(
+                ${t.vars.palette.action.selectedOpacity} + 
+                ${t.vars.palette.action.hoverOpacity}
+              ))`
+    : alpha(
+        t.palette.primary.main,
+        t.palette.action.selectedOpacity + t.palette.action.hoverOpacity,
+      );
+
+  const pinnedHoverBackground = t.vars
+    ? hoverBackground
+    : blend(pinnedBackground, hoverBackground, hoverOpacity);
+  const pinnedSelectedBackground = t.vars
+    ? selectedBackground
+    : blend(pinnedBackground, selectedBackground, selectedOpacity);
+  const pinnedSelectedHoverBackground = t.vars
+    ? hoverBackground
+    : blend(pinnedSelectedBackground, hoverBackground, hoverOpacity);
+
+  const selectedStyles = {
+    backgroundColor: selectedBackground,
+    '&:hover': {
+      backgroundColor: selectedHoverBackground,
+      // Reset on touch devices, it doesn't add specificity
+      '@media (hover: none)': {
+        backgroundColor: selectedBackground,
+      },
+    },
+  };
 
   const gridStyle: CSSInterpolation = {
     '--unstable_DataGrid-radius': typeof radius === 'number' ? `${radius}px` : radius,
-    '--unstable_DataGrid-headWeight': theme.typography.fontWeightMedium,
-    '--unstable_DataGrid-overlayBackground': theme.vars
-      ? `rgba(${theme.vars.palette.background.defaultChannel} / ${theme.vars.palette.action.disabledOpacity})`
-      : alpha(theme.palette.background.default, theme.palette.action.disabledOpacity),
+    '--unstable_DataGrid-headWeight': t.typography.fontWeightMedium,
+    '--unstable_DataGrid-overlayBackground': overlayBackground,
+
+    '--DataGrid-containerBackground': containerBackground,
+    '--DataGrid-pinnedBackground': pinnedBackground,
+    '--DataGrid-rowBorderColor': borderColor,
```

Comment:
One side border for [spanned columns](https://deploy-preview-10059--material-ui-x.netlify.app/x/react-data-grid/column-spanning/) seems to not have picked the right color.
![image](https://github.com/mui/mui-x/assets/12609561/962edf3a-7735-4ba8-805f-90f189e10089)